### ISSUE TITLE
[hmac,dv] Analyze unreached feature coverage items

### ISSUE THREAD

**Issue by martin-velay** [lowRISC/opentitan#23594]
### Description

The current feature coverage shows that some stuffs are not exercised:

- [x] msg_len_cg: long messages, and some others auto lengths (closed by https://github.com/lowRISC/opentitan/pull/23864 and https://github.com/lowRISC/opentitan/pull/23901)
- [x] pending_req_on_rst_cg.cp_req_pending (closed by https://github.com/lowRISC/opentitan/pull/25843)
- [x] sha2_invalid, sha2_none and key_invalid/none (closed by https://github.com/lowRISC/opentitan/pull/23843)
- [x] tl_errors_cg.cp_mem_byte_access_err/cp_mem_ro_err

**Comment by andreaskurth** [comment]
@martin-velay: Should we move this to V3 and M7?

**Comment by martin-velay** [comment]
@andreaskurth: I think we should keep it in M5 as it's quite important to ensure at least that, those items are covered when all the ongoing related PR will have been merged:
- msg_len_cg: long messages, and some others auto lengths
- sha2_invalid, sha2_none and key_invalid/none

**Comment by martin-velay** [comment]
@rswarbrick 
Could you tell me if you have any idea how to cover the 2 remaining items? As those covergroups are coming from the common DV framework.

**Comment by rswarbrick** [comment]
I misunderstood this message at first and thought you were talking about the remaining coverage points from your message on 26/7.

On those coverage points items:
- I think I agree with what you said, but...
- Looking at last night's run, `sha2_invalid`, `sha2_none` (in the `digest_size` coverpoint of `cfg_cg`) are hit.
- Similarly, `key_invalid` and `key_none` (in the `key_length` coverpoint of `cfg_cg`) are hit.
- The `msg_len_cg` results should be better after the fix that you merged this morning, thanks.
- But I notice that we haven't seen any hits in `msg_len_upper_cp` for some reason. Does that mean we never give a massive message length with a nonzero upper word? I don't think this has been fixed yet?

(I'll respond to the other items in a minute)

**Comment by rswarbrick** [comment]
For `pending_req_on_rst_cg.cp_req_pending`:

I think this wants to see a reset when there is a pending request (so `pending_a_req` in `tl_monitor` is nonempty). This gets pushed to when we see a request on the A side (in `check_a_channel`) and then popped when we see the response (in `check_d_channel`).

The weird bit is that I think we're currently actively avoiding doing a reset in the middle of a TL transaction. This is avoided by the code in `wait_to_issue_reset` in `cip_base_vseq`. I believe that @matutem is the expert on this code. Matute, how would you expect this coverpoint to be hit? (And is there something that we're going to need to change?)

**Comment by rswarbrick** [comment]
For `tl_errors_cg.cp_mem_byte_access_err`, I just had to do some grepping to figure out what's going on and I found `cip_base_scoreboard.sv`. If you look at line 105, you'll see that we might essentially disable the coverpoint if the call to `get_all_mem_attrs` says that the RAL model doesn't support this error type.

Looking at the code (in `cip_base_pkg.sv`), it looks like this error type is considered supported if one of the memories backing the register block is read-only and has no partial write support. I guess that must be what's happening here.

Oddly, the flag in question seems to be set at runtime in `cip_base_scoreboard::is_tl_access_allowed` (line 549 of `cip_base_scoreboard.sv`), when we see a partial write.

Debugging the code in question:
- Could you add a print message to the `cip_base_pkg.sv` code to figure out which memory we think should allow this event?
- Do you think the event *should* be possible?
- And does the code in `cip_base_pkg.sv` look right to you? I'm slightly surprised that we're only interested in read-only blocks and I notice that this change appeared in March 2022 with b4f912be806 to "fix coverage holes". It looks a lot to me like this should have been "!=", but it's possible that I've misunderstood something.

**Comment by rswarbrick** [comment]
For `tl_errors_cg.cp_mem_ro_err`, I think we're in roughly the same place. Here, the coverpoint is only enabled if we have a read-only memory (`cip_base_scoreboard`, line 113). To trigger it, we need to see a write to a read-only memory (`cip_base_scoreboard`, line 561).

I think the code for this was added with 7c33aa4c804 in 2021. The sequence that's supposed to be causing the error is in `hw/dv/sv/cip_lib/seq_lib/cip_base_vseq__tl_errors.svh` (the `tl_write_ro_mem_err` task, called on line 257 if there is a read-only memory).

Would you mind picking up the search from here? The first question is a bit like the previous note. Do we actually have a read-only memory, or is something labelled incorrectly?

**Comment by martin-velay** [comment]
> I misunderstood this message at first and thought you were talking about the remaining coverage points from your message on 26/7.
> 
> On those coverage points items:
> 
> * I think I agree with what you said, but...
> * Looking at last night's run, `sha2_invalid`, `sha2_none` (in the `digest_size` coverpoint of `cfg_cg`) are hit.
> * Similarly, `key_invalid` and `key_none` (in the `key_length` coverpoint of `cfg_cg`) are hit.
> * The `msg_len_cg` results should be better after the fix that you merged this morning, thanks.
> * But I notice that we haven't seen any hits in `msg_len_upper_cp` for some reason. Does that mean we never give a massive message length with a nonzero upper word? I don't think this has been fixed yet?
> 
> (I'll respond to the other items in a minute)

About the upper part of the message counter I tried to make it yesterday without success as the simulation will take ages, but this morning Pirmin gave me a trick to cheat a bit. I'll implement this to cover that one.

**Comment by andreaskurth** [comment]
Just discussed in triage: We reached the V2S coverage goals for HMAC, so moving this to M7 as part of V3

**Comment by martin-velay** [comment]
> For `pending_req_on_rst_cg.cp_req_pending`:
> 
> I think this wants to see a reset when there is a pending request (so `pending_a_req` in `tl_monitor` is nonempty). This gets pushed to when we see a request on the A side (in `check_a_channel`) and then popped when we see the response (in `check_d_channel`).
> 
> The weird bit is that I think we're currently actively avoiding doing a reset in the middle of a TL transaction. This is avoided by the code in `wait_to_issue_reset` in `cip_base_vseq`. I believe that [@matutem](https://github.com/matutem) is the expert on this code. Matute, how would you expect this coverpoint to be hit? (And is there something that we're going to need to change?)

I have checked half of the IPs, and found out that most of them have the same hole, but [i2c](https://reports.opentitan.org/hw/ip/i2c/dv/latest/cov_report/grp23.html) and [adc_ctrl](https://reports.opentitan.org/hw/ip/adc_ctrl/dv/latest/cov_report/grp14.html). I have no idea yet how they've managed this.

**Comment by martin-velay** [comment]
@rswarbrick, regarding this comment https://github.com/lowRISC/opentitan/issues/23594#issuecomment-2202787462. At the end your are saying that you don't understand why in this PR https://github.com/lowRISC/opentitan/pull/10595 the `dv_mem.get_access() == "RO"`. I have to say, me neither.

@weicaiyang / @matutem do you remember the reason?

**Comment by martin-velay** [comment]
Regarding `tl_agent_pkg::pending_req_on_rst_cg.cp_req_pending` hole, I have parsed the latest regression results for each IPs. Some are covered via Xcelium and I cannot see the result and some others are covered by formal. Here is the list of relevant blocks and their status about this coverpoint:

```
adc_ctrl        -> covered
aon_timer       -> not
edn             -> not
gpio            -> not
hmac            -> not
i2c             -> covered
keymgr          -> not
kmac            -> not
otp_ctrl        -> covered
rom_ctrl        -> not
rv_timer        -> covered
spi_device      -> not
sram_ctrl       -> partially
sysrst_ctrl     -> covered
uart            -> not
usbde           -> not
```

Now I'll try to see how some managed to have this covered

**Comment by rswarbrick** [comment]
About the fact that one has to set a `goal`: if I have a coverpoint with weight zero, the report might still show its coverage as being low. The weight doesn't change whether the coverpoint gets hit: it just tells the tool that we don't care :-) As a result, I'd hope that the "next level up" would not take the low number into account.

**Comment by martin-velay** [comment]
@rswarbrick
> I'd hope that the "next level up" would not take the low number into account

After studying more the numbers you're right, this "hole" is flagged at one level but but taken into account at the level up. I have been confused by the score displayed on the 2nde line on the screenshot below (purple rectangle), which says 86.67% covered. But looking at the overall group coverage, it says that 683 over 684 items are covered (blue rectangle). This last remaining item is referring to the `pending_req_on_rst_cg`.

![Image](https://github.com/user-attachments/assets/fbeba80b-a45e-4bf8-9ea5-b5ef51a8c801)

So I'll drop my last changes and sorry for this noise.

**Comment by martin-velay** [comment]
I wonder if we really need to cover all the TL agent items for each IPs as we are just reusing this block which is supposed to be 100% covered and verified at it's own level. According to me this missing `pending_req_on_rst_cg` can can be simply excluded.

What do you think @rswarbrick ?

**Comment by rswarbrick** [comment]
I agree that trying to cover every instance is probably more work than the benefit it gives. But we'd probably need a sensible story about why something could be waived without losing confidence in the chip. This ties in nicely with the vplan work you're doing: maybe we have to do the "pointless work" for a short time until we can switch to something structured like we have talked about.

**Comment by martin-velay** [comment]
The last remaining item `pending_req_on_rst_cg.cp_req_pending` is now covered, probably thanks to this PR https://github.com/lowRISC/opentitan/pull/25843
